### ISSUE TITLE
`yarn` binary exists in multiple repositories

### ISSUE THREAD

**Issue by vladdeSV** [teaxyz/cli#383]
Yarn did a split some time ago, where multiple versions exists in two repositories

- [yarnpkg/yarn](https://github.com/yarnpkg/yarn) (classic, v1)
- [yarnpkg/berry](https://github.com/yarnpkg/berry) (current, v2;v3)

How should this be handled?

**Comment by vladdeSV** [comment]
Yarn Classic is well maintained and will most likely never release anything else than v1, so you could possibly do some sort of version comparison; `yarn^1` would use classic Yarn, and `yarn` would default to latest `v3`.

However, when you link to a GitHub tags page for versions, Yarn Classic could release a v2 there and there would be name and version collisions.

**Comment by jhheider** [comment]
I would say those are probably two different packages: `yarnpkg.com` and `yarnpkg.com/berry`. Easy to do it that way, since they have separate GitHub repos.

**Comment by vladdeSV** [comment]
With tea magic, the issue is which repository should be used when calling `yarn`?

**Comment by jhheider** [comment]
My gut says `yarnpkg.com` will get picked up before `yarnpkg.com/berry`.  Assuming the preferred default is the latter, I'd do

```sh
cd {{prefix}}/bin
ln -s yarn yarn1

provides:
  - bin/yarn1
```

In the former. Then you can get it using the magic command `yarn1` or you can inject the package specifically using any of the methods and invoke it as `yarn`, while magic will give you berry.

**Comment by vladdeSV** [comment]
It is my opinion that versioning the binary is not the way to go.

In this example it is about Yarn, but a similar scenario could just as well happen with the decentralized approach of pantries in the future? (If I understood decentralized pantries correctly)

**Comment by jhheider** [comment]
This versioning would only apply to pure-magic invocations. You can still always do `yarn^1 install` or `yarn'>=2' install` as needed, or be specific with `tea +yarnpkg.com/berry yarn install`.

**Comment by mxcl** [comment]
Without significant changes to how tea (currently) works, we need to use the same package.yml.

Possibly in the future these should be separate yamls, but we can't do that RN without failing to find versions based on whichever yaml is picked non-determistically.

**Comment by jhheider** [comment]
> Without significant changes to how tea (currently) works, we need to use the same package.yml.

For different github repos? How?

**Comment by vladdeSV** [comment]
I am interested in how tea magick would work with a decentralized pantry in the future. If I create my own pantry with my own binary `yarn`, how would it be differentiated from that belonging tro `+yarnpkg.com`?

**Comment by mxcl** [comment]
> > Without significant changes to how tea (currently) works, we need to use the same package.yml.
> 
> For different github repos? How?

My thinking has been something like:

```yaml
distributable:
  ^1: URL
  ^2: OTHER-URL
```

I expect there’s a better way to do this out there, but I haven't come up with it yet.

**Comment by mxcl** [comment]
> I am interested in how tea magick would work with a decentralized pantry in the future. If I create my own pantry with my own binary yarn, how would it be differentiated from that belonging tro +yarnpkg.com?
> 
> (bit off-topic, not related directly to this issue)

Yes I have been thinking about this too. We would need some kind of ping step that adds an entry to some datastore.

So ping, then add a note that `foo` can be fetched from this other source.

Users would obv. have to confirm they are cool with the other source before we would auto-fetch though.
